### PR TITLE
refactor: proper EEXIST detection

### PR DESCRIPTION
**Pull request by KSXGitHub**

### CODE REVIEW

**Comment by KSXGitHub** [comment]
@nachoaldamav Just to be sure, `err.code === 'ENOENT'` is also true on Windows, right?

**Comment by nachoaldamav** [comment]
@KSXGitHub Yes, Windows throws the error 
```pwsh
ENOENT: no such file or directory, open 'E:\nonexistent_file.txt'
```
The only issue I can think of is this #7554, the pnpm reflink package does not use node error codes, but OS codes instead, how should we handle this?

**Comment by KSXGitHub** [comment]
> The only issue I can think of is this #7554, the pnpm reflink package does not use node error codes, but OS codes instead, how should we handle this?

I guess we could fork https://github.com/pnpm/reflink? I cringed a little bit when I saw its [main.rs](https://github.com/pnpm/reflink/blob/808dbb57f2eacb186b567b4c85903246fd147025/src/lib.rs) was using `.unwrap()`. @zkochan What do you think?

**Comment by zkochan** [comment]
Why do you need to fork reflink if it is already in the pnpm org?

**Comment by nachoaldamav** [comment]
@KSXGitHub You mean [nachoaldamav/CopyOnWrite](https://github.com/nachoaldamav/CopyOnWrite)?

**Comment by KSXGitHub** [comment]
> Why do you need to fork reflink if it is already in the pnpm org?

I did a quick search and saw that the TypeScript code uses `@reflink/reflink` so I was mistaken.

File: fs/indexed-pkg-importer/src/index.ts [hunk 1]
```diff
@@ -136,11 +136,10 @@ function createCloneFunction (): CloneFunction {
       try {
         reflinkFileSync(fr, to)
       } catch (err: unknown) {
-        assert(util.types.isNativeError(err))
         // If the file already exists, then we just proceed.
         // This will probably only happen if the package's index file contains the same file twice.
         // For instance: { "index.js": "hash", "./index.js": "hash" }
-        if (!err.message.startsWith('File exists') && !err.message.includes('-2147024816')) throw err
+        if (!util.types.isNativeError(err) || !('code' in err) || err.code !== 'ENOENT') throw err
```
Comment (KSXGitHub) [inline fs/indexed-pkg-importer/src/index.ts]: I'm dumb.

```suggestion
        if (!util.types.isNativeError(err) || !('code' in err) || err.code !== 'EEXIST') throw err
```

Comment (nachoaldamav) [reply fs/indexed-pkg-importer/src/index.ts]: Could we add the OS error code for now for reflinks? it's an edgecase but as we already had the same issue before

Comment (KSXGitHub) [reply fs/indexed-pkg-importer/src/index.ts]: It's just a refactor. There's no rush to merge this. The old code still works just fine.

**Review by zkochan — APPROVED** [review]

**Comment by zkochan** [comment]
a changeset is needed

**Pull request merged**